### PR TITLE
dialects: Add basic MPI dialect

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR aims to add very basic MPI dialect operations:

 - `Send` and `ISend`
 - `Recv` and `IRecv`
 - `Test`
 - `Wait`

Declared as out of scope for now in this dialect:

 - Explicit handling of `MPI_Comm`
 - Receiving into an existing `memref`
 - Handling function error codes (nonzero return values)
 - Non compile time constant `tag`s
 - Non compile time known sizes and types for send/recv

### CODE REVIEW

**Review by webmiche — COMMENTED** [review]
Nice, this looks like a fairly mature MPI abstraction already. I didn't notice anything that particularly shocked me, if you have specific questions/discussions, feel free to reach out.

Also, I suggest trying to lower this sooner rather than later as I myself sometimes had to notice issues in the design only when trying to actually get it to run vertically.

File: xdsl/dialects/mpi.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+class StatusTypeField(Enum):
+    """
+    This enum lists all fields in the MPI_Status struct
+    """
+    MPI_SOURCE = 'MPI_SOURCE'
+    MPI_TAG = 'MPI_TAG'
+    MPI_ERROR = 'MPI_ERROR'
+
+
+class MPIBaseOp(Operation, ABC):
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/mpi.py:47]: While working with the RIscv_ssa dialect, I often found myself wanting a way to check if an op belongs to a specific dialect. I would propose that all operations of a specific dialect subclassing from a common abstract base class is a good solution for that problem.

But then I don't know how common that problem really is.

**Comment by webmiche** [comment]
Also, this to me starts to look a bit "upward facing" with encoding things in the memref. I think it might be worth being careful to not implement it too nice and make lowering painful afterward. Introducing the upward facing IR can still be done later, if needed.

**Comment by AntonLydike** [comment]
> Also, this to me starts to look a bit "upward facing" with encoding things in the memref. I think it might be worth being careful to not implement it too nice and make lowering painful afterward. Introducing the upward facing IR can still be done later, if needed.

That's a very valid concern. I have very limited experience with this, do you have some ideas on how we can improve this?

I don't think we want to use raw pointer types, right?

File: xdsl/dialects/mpi.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+from xdsl.ir import OpResult, ParametrizedAttribute, Dialect, Operation
+from xdsl.irdl import (Operand, Annotated, irdl_op_definition,
+                       irdl_attr_definition, OptOpAttr, OpAttr)
+from xdsl.dialects.builtin import (IntegerType, Signedness, IntegerAttr,
+                                   AnyFloatAttr, AnyIntegerAttr, StringAttr)
+from xdsl.dialects.memref import MemRefType, Alloc
+
+t_uint32: IntegerType = IntegerType.from_width(32, Signedness.UNSIGNED)
+t_int: IntegerType = IntegerType.from_width(32, Signedness.SIGNED)
+t_bool: IntegerAttr = IntegerType.from_width(1, Signedness.SIGNLESS)
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/mpi.py]: Fixed!

This bot is pretty awesome!

File: xdsl/dialects/mpi.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                         result_types=[dtype, RequestType()])
+
+
+@irdl_op_definition
+class Recv(MPIBaseOp):
+    """
+    This wraps the MPI_Recv function (blocking receive).
+    https://www.mpich.org/static/docs/v4.1/www3/MPI_Recv.html
+
+    ## The MPI_Irecv Function Docs:
```
Comment (wence-) [inline xdsl/dialects/mpi.py:197]:
```suggestion
    ## The MPI_Recv Function Docs:
```

File: xdsl/dialects/mpi.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+    ## Our Abstractions:
+
+        - We bundle buf, count and datatype into the type definition and use `memref`
+        - We assume this type information is compile-time known
+        - We assume tag is compile-time known
+        - We omit the possibility of using multiple communicators
+    """
+
+    name = "mpi.irecv"
```
Comment (wence-) [inline xdsl/dialects/mpi.py:218]:
```suggestion
    name = "mpi.recv"
```

Comment (AntonLydike) [reply xdsl/dialects/mpi.py:218]: Good catch! Thanks!

**Comment by webmiche** [comment]
IMO, you might want to go back to only send and recv and then get a vertical prototype running. This file can have everything hard-coded (like buffer allocation or sizes) as long as it really works. Then, start parametrizing stuff, actually add the size and allocation and so on, always making sure that the vertical prototype still works. 

Also, you can just add a dummy operation for all the hardcoded stuff, if you want. Then you dont have to hack as much and continue to move things out during your parametrization process.

The core decision that needs to be done before being able to run the prototype is the level you want to link against MPI. I feel both C and LLVM IR will work, but I would personally suggest generating C as a first step. It has more language features and getting the actual binary might be easier...

**Review by georgebisbas — APPROVED** [review]

File: xdsl/dialects/mpi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (math-fehr) [reply xdsl/dialects/mpi.py:47]: Yeah, the solutions are either to check for the dialect name in the operation string, or doing this.
I'm fine with any!

File: xdsl/dialects/mpi.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        - We assume that tag is compile-time constant
+        - We omit the possibility of using multiple communicators
+    """
+
+    name = 'mpi.send'
+
+    buffer: Annotated[Operand, MemRefType[AnyNumericAttr]]
+    dest: Annotated[Operand, t_int]
+
+    tag: OptOpAttr[IntegerAttr]
```
Comment (math-fehr) [inline xdsl/dialects/mpi.py:136]:
```suggestion
    tag: OptOpAttr[AnyIntegerAttr]
```

File: xdsl/dialects/mpi.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        - We assume this type information is compile-time known
+        - We assume tag is compile-time known
+        - We omit the possibility of using multiple communicators
+    """
+
+    name = "mpi.recv"
+
+    source: Annotated[Operand, t_int]
+
+    tag: OptOpAttr[IntegerAttr]
```
Comment (math-fehr) [inline xdsl/dialects/mpi.py:222]:
```suggestion
    tag: OptOpAttr[AnyIntegerAttr]
```

File: xdsl/dialects/mpi.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        - We assume that tag is compile-time constant
+        - We omit the possibility of using multiple communicators
+    """
+
+    name = 'mpi.isend'
+
+    buffer: Annotated[Operand, MemRefType[AnyNumericAttr]]
+    dest: Annotated[Operand, t_int]
+
+    tag: OptOpAttr[IntegerAttr]
```
Comment (math-fehr) [inline xdsl/dialects/mpi.py:95]:
```suggestion
    tag: OptOpAttr[AnyIntegerAttr]
```

File: xdsl/dialects/mpi.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        - We assume this type information is compile-time known
+        - We assume tag is compile-time known
+        - We omit the possibility of using multiple communicators
+    """
+
+    name = "mpi.irecv"
+
+    source: Annotated[Operand, t_int]
+
+    tag: OptOpAttr[IntegerAttr]
```
Comment (math-fehr) [inline xdsl/dialects/mpi.py:176]:
```suggestion
    tag: OptOpAttr[AnyIntegerAttr]
```

**Review by math-fehr — APPROVED** [review]
Perfect! I just added suggestions on some pyright issues, but otherwise it's good to merge!

**Review by webmiche — APPROVED** [review]

**Comment by webmiche** [comment]
I still don't know if this design can be lowered, but that is probably something for a different PR :)

**Pull request merged**